### PR TITLE
transformer: perform direct access to array when the index is known as safe

### PR DESCRIPTION
**Pull request by thomas-mangin**
This PR removes boundary checks for array access when using integer literals such as is found in the binary module (and is common in Go).
```
[inline]
pub fn little_endian_u32(b []byte) u32 {
	_ = b[3] // bounds check
	return u32(b[0]) | (u32(b[1]) << u32(8)) | (u32(b[2]) << u32(16)) | (u32(b[3]) << u32(24))
}
```

The code is also analysing `assert` which may be a preferable way to perform the check, the following code may be considered more idiomatic:
```
[inline]
pub fn little_endian_u32(b []byte) u32 {
	assert b.len >= 4
	return u32(b[0]) | (u32(b[1]) << u32(8)) | (u32(b[2]) << u32(16)) | (u32(b[3]) << u32(24))
}
```

The AST tree is analysed during the transformer pass, where `IndexExpr` nodes are annotated with a new `is_direct` boolean, indicating that the array index can be safely generated, similarly to when the function is annotated with `[direct_array_access]`.

The c_gen code check was changed to handle this annotation. The following cases are considered valid for this optimisation:
1. the array was created with a `len` larger than the index requested
2. the array was previously accessed with a higher value which would have reported the issue already
3. the array is the result of a range (such as `a := b[0..5]`) and the offset was safe in the original array (any further access to the created array will not update maximum access to the parent)

Current limitations:
 * any function using break/continue or goto/label stopped from being optimised as soon as the relevant AST nodes are found as the code can not be ensured to be sequential
 * `enum` and `ident` indexes are not optimised (`enum` could probably be looked up)
 * the init and inc of for are not analysed (ATM)

### CODE REVIEW

File: do [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+[ -f test ] && rm test
+[ -f vt ] && rm vt
+echo
+echo "making V"
+./v -d debug_bounds_checking -o vt cmd/v
+echo
+echo "compiling test"
+./vt test.v -printfn main__main
+#echo "testing"
+#[ -f ./test ] && ./test
```
Comment (spytheman) [inline do:13]: this should be removed from the PR imho

File: test.v [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	safe = a[int(10)]
+
+	assert a.len >= 19
+	a[17] = a[18]
+
+	// safe a[Index.one] not Index.two
+	a[Index.one] = a[Index.two]
+
+	println('all done')
+}
```
Comment (spytheman) [inline test.v:86]: This should be moved as a _test.v file under `vlib/v/tests/`, or if you prefer under `vlib/v/tests/testdata/`.
The difference is that files under `testdata/` are ignored by `v test .`, so if they are source code, they need a special _test.v file that knows how to run them (with what options etc). The shell script can be translated to such a _test.v file actually.

File: vlib/v/transformer/transformer.v [hunk 3]
```diff
@@ -4,46 +4,275 @@ import v.pref
...
+
+[if debug_bounds_checking ?]
+fn debug_bounds_checking(str string) {
+	println(str)
+}
+
+// IndexState is used to attempt to remove boundary check for integer literal array access
+// a[1]. The code will perform a direct access when the length of the array is known
+// to be bigger than the index when:
+// (1) the array was created with a larger len
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:20]: mut arrays have methods that change .len, like .delete or .trim

File: vlib/v/transformer/transformer.v [hunk 4]
```diff
@@ -4,46 +4,275 @@ import v.pref
...
+// all previous information is dropped
+[if !no_bounds_checking]
+fn (mut i IndexState) clean_indent() {
+	mut kvs := []KeyVal{cap: i.max_index.len}
+	for k, v in i.max_index {
+		kvs << KeyVal{k, v}
+		i.max_index.delete(k)
+	}
+	i.saved_disabled << i.disabled
+	i.saved_key_vals << kvs
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:70]: > all previous information is dropped

why do you save it here?

Comment (spytheman) [reply vlib/v/transformer/transformer.v:70]: because of anonymous functions?

File: vlib/v/transformer/transformer.v [hunk 5]
```diff
@@ -4,46 +4,275 @@ import v.pref
...
+	i.disabled = i.saved_disabled.pop()
+}
+
+// restoring the data as it was before the if/for/unsafe block
+[if !no_bounds_checking]
+pub fn (mut i IndexState) balanced() {
+	if i.level != 0 {
+		panic('unbalanced index indent/unindent')
+	}
+}
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:113]: the behavior does not match the comment

File: vlib/v/transformer/transformer.v [hunk 6]
```diff
@@ -4,46 +4,275 @@ import v.pref
...
-pub fn (t Transformer) stmt(mut node ast.Stmt) {
+pub fn (mut t Transformer) find_new_array_len(node ast.AssignStmt) {
+	// looking for, array := []type{len:int}
+	mut right := node.right[0]
+	if mut right is ast.ArrayInit {
+		mut left := node.left[0]
+		if mut left is ast.Ident {
+			// as we do not need to check any value under the setup len
+			if !right.has_len {
+				t.index.safe_access(left.name, -1)
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:152]: I am curious, why `-1`, instead of `0`?
Below, you use `t.index.safe_access(left.name, len)`, when len is set explicitly, so I assume that the second param is in a 1:1 relationship to `x` in `[]byte{len:x}`?

File: vlib/v/transformer/transformer.v [hunk 7]
```diff
@@ -4,46 +4,275 @@ import v.pref
...
+				len = value.val.int()
+			}
+
+			t.index.safe_access(left.name, len)
+		}
+	}
+}
+
+pub fn (mut t Transformer) find_assert_len(node ast.InfixExpr) {
+	right := node.right
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:169]: this will only work for `assert a.len == 5`, but not for `assert 5 == a.len`

File: vlib/v/transformer/transformer.v [hunk 8]
```diff
@@ -4,46 +4,275 @@ import v.pref
...
+		}
 		ast.AssignStmt {
-			for mut right in node.right {
-				right = t.expr(right)
+			t.find_new_array_len(node)
+			for right in node.right {
+				t.expr(right)
+			}
+			for left in node.left {
+				t.expr(left)
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:261]: this changes the semantics of the original:
```v
for mut right in node.right {
	right = t.expr(right)
```
, means that the right value is transformed, and then the result replaces the original right value, thus stuff like:
`a := 'a' + 'b'` => `a := 'ab'`

With the new code, that replacement will not happen afaik.

File: vlib/v/transformer/transformer.v [hunk 9]
```diff
@@ -74,24 +303,60 @@ pub fn (t Transformer) stmt(mut node ast.Stmt) {
...
 				cond: t.expr(node.cond)
 			}
-			if node.cond is ast.BoolLiteral && !(node.cond as ast.BoolLiteral).val { // for false { ... } should be eleminated
-				node = &ast.EmptyStmt{}
+			match node.cond {
+				ast.BoolLiteral {
+					if !(node.cond as ast.BoolLiteral).val { // for false { ... } should be eleminated
+						node = &ast.EmptyStmt{}
+					}
+				}
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:338]: that is unrelated to the main topic of the PR

Comment (spytheman) [reply vlib/v/transformer/transformer.v:338]: oh, it was already there before, you just refactored it - my mistake, sorry.

File: vlib/v/transformer/transformer.v [hunk 10]
```diff
@@ -74,24 +303,60 @@ pub fn (t Transformer) stmt(mut node ast.Stmt) {
...
 			}
 		}
 		ast.GlobalDecl {}
-		ast.GotoLabel {}
-		ast.GotoStmt {}
+		ast.GotoLabel {
+			// we can not rely on sequential scanning and need to cancel all index optimisation
+			t.index.disabled = true
```
Comment (spytheman) [inline vlib/v/transformer/transformer.v:354]: goto labels by themselves are nops; I do not think they should be taken into account

File: do (hunk 1 above)
Comment (thomas-mangin) [reply do:13]: I agree, this is in a separate commit as I did not know where to place it.

File: test.v (hunk 2 above)
Comment (thomas-mangin) [reply test.v:86]: All code will always run with this patch, the issue could be excessive check removal which is what this code was written to try to notice.

File: vlib/v/transformer/transformer.v (hunk 3 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:20]: Thank you - indeed. I overlooked this case, will update the patch to ignore mut array.

File: vlib/v/transformer/transformer.v (hunk 4 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:70]: We can not analyse if/for paths (as the for or if block may not run) so we save the state before an if/for loop, continue to update index in the branch, but when we leave we need to undo all updates. For example:
```
_ = a[5]
if false {
 _ = a[7]
 _ = a[6] // safe within the if
}
_ = a[6] // unsafe if did not run
```
So we want to call indent after the `false` Expr has been analysed and unindent when we leave the if block. This is where my unperfect understanding of the AST may lead to excessive check removal and where I would appreciate the most a review.

File: vlib/v/transformer/transformer.v (hunk 5 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:113]: Happy to remove this, this is really debugging code.

File: vlib/v/transformer/transformer.v (hunk 6 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:152]: if `len == 0`, it means that `a[0]` is invalid ... so accesses are one less than the len. the number saved is the last valid array access seen `a[5]` gives 5. I will update the comments to be clearer.

File: vlib/v/transformer/transformer.v (hunk 7 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:169]: In that case, it will not work if there is assert `a.len == 5 && b.len == 6`. Happy to provide the other order too but really this is here as a guard and the guard could have a fixed for .. Happy to improve the code that way if it is what you think should be done. I was not even sure if the assert should be added at all :-)

File: vlib/v/transformer/transformer.v (hunk 8 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:261]: well spotted - thanks - will fix.

File: vlib/v/transformer/transformer.v (hunk 10 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:354]: I was being overly careful here - I agree - will change.

File: do (hunk 1 above)
Comment (thomas-mangin) [reply do:13]: I agree, I added all the code so one could see how I used the test code. It should be removed.

File: test.v (hunk 2 above)
Comment (thomas-mangin) [reply test.v:86]: ok, I will write some test code that checks that the C output is as should.

File: vlib/v/transformer/transformer.v (hunk 7 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:169]: On second thought, it may be better staying as it to not see things `assert false && a.len == 5`, which would force a full analysis of the infix expression ...

File: vlib/v/transformer/transformer.v (hunk 8 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:261]: This is a subtility that may gain from documentation in the doc ..

File: vlib/v/transformer/transformer.v (hunk 6 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:166]: comment updated, let me know if it is now clearer.

File: vlib/v/transformer/transformer.v (hunk 4 above)
Comment (spytheman) [reply vlib/v/transformer/transformer.v:87]: the comment at the top of the function is misleading then

Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:87]: I rewrote the comment section, can you please let me know if you believe it is now clearer please?

File: vlib/v/transformer/transformer.v (hunk 6 above)
Comment (spytheman) [reply vlib/v/transformer/transformer.v:166]: > if len == 0, it means that a[0] is invalid

that is the case for `a := []byte{}` - .len is 0, and `a[0]` is not valid. `a := []byte{len:1}` - .len is 1, and `a[0]` is valid, but `a[1]` is not, etc

File: vlib/v/transformer/transformer.v (hunk 7 above)
Comment (spytheman) [reply vlib/v/transformer/transformer.v:183]: I do not want a full analysis (which while nice, is much more complex), just symmetry, so people that do prefer the different code style (C programmers) do not get affected disproportionately.

File: vlib/v/transformer/transformer.v (hunk 6 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:166]: Thank you ...
```
--- a/vlib/v/transformer/transformer.v
+++ b/vlib/v/transformer/transformer.v
@@ -171,7 +171,7 @@ pub fn (mut t Transformer) find_new_array_len(node ast.AssignStmt) {
 
                        value := right.len_expr
                        if value is ast.IntegerLiteral {
-                               len = value.val.int()
+                               len = value.val.int() + 1
                        }
 
                        t.index.safe_access(left.name, len)
```

File: vlib/v/transformer/transformer.v (hunk 7 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:183]: sure. Adding it now.

File: vlib/v/transformer/transformer.v (hunk 4 above)
Comment (spytheman) [reply vlib/v/transformer/transformer.v:87]: no, it is not - it is still confusing to me:
```
// all previous information is dropped
```
and later:
```v
i.saved_disabled << i.disabled
```
are contradictory - the comment should not be there imho

Comment (spytheman) [reply vlib/v/transformer/transformer.v:87]: `dropped` to me implies losing/removing something, because you do not care about it anymore.

Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:87]: Fixing things so please give me a little bit of time before continuing the review :-)

Comment (spytheman) [reply vlib/v/transformer/transformer.v:87]: NB: I do not object to the mechanism, but to the comment line that creates the confusion.

Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:87]: Found an unrelated bug - so wanting to fix it before you continue looking

File: vlib/v/transformer/transformer.v (hunk 6 above)
Comment (spytheman) [reply vlib/v/transformer/transformer.v:166]: imho, it would be better (more maintainable), if that calculation is encapsulated inside safe_access, so that calls to it (which are many) can be clearer

File: test.v (hunk 2 above)
Comment (thomas-mangin) [reply test.v]: totally rewrote comment with .. some bad English I will correct later on .. sigh:

 // indent is used for when encountering new code blocks (if, for and functions)
 // The code analysis needs to take into consideration blocks of code which
 // may not run at runtime (if/for) and therefore even if a new maximum for an
 // index access is found on an if branch it can not be used within the parent
 // code. The same is true with for blocks. indent() snapshot the current state,
 // to allow restoration with unindent()
 // Also within a function, analysis must be `disabled` when goto or break are
 // encountered as the code flow is then not lineear, and only restart when a
 // new function analysis is started.

File: vlib/v/transformer/transformer.v (hunk 4 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:93]: totally rewrote comment with .. some bad English I will correct later on .. sigh:

// indent is used for when encountering new code blocks (if, for and functions)
// The code analysis needs to take into consideration blocks of code which
// may not run at runtime (if/for) and therefore even if a new maximum for an
// index access is found on an if branch it can not be used within the parent
// code. The same is true with for blocks. indent() snapshot the current state,
// to allow restoration with unindent()
// Also within a function, analysis must be disabled when goto or break are
// encountered as the code flow is then not lineear, and only restart when a
// new function analysis is started.

File: vlib/v/transformer/transformer.v (hunk 6 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:155]: safe_access API is to let the caller know if access to an index is safe or not at this point in the code. The logic of determining what was asked (parsing the AST and finding the index) is already into self-contained Transformer functions (find_new_array_len, find_assert_len, ..) so that this can be abstracted from the main AST parsing. IMHO, getting more into IndexState would require making it AST aware which would break the API boundary in a worse way. What do you think?

File: vlib/v/transformer/transformer.v (hunk 7 above)
Comment (thomas-mangin) [reply vlib/v/transformer/transformer.v:176]: done.

File: test.v (hunk 2 above)
Comment (thomas-mangin) [reply test.v]: test available
